duckduckgo/BrowserServicesKit
# VPN server failure detection recovery

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/72649045549333/1206669340827392/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2779
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2657
What kind of version bump will this require?: Major

**Description**:

- Use the WireGuard handshakes to determine when a connection fails by checking that the last handshake time is less than 300 seconds (just over two handshakes).
- When a failure is detected the client will call /register with server set to the name of the current (failed) server and mode set to failureRecovery. If the server is not in the "active" status on the controller a new registration will be returned otherwise the previous data will be returned (this data should be checked against the current tunnel data including allowedIPs).

**Steps to test this PR**:
1. Launch the VPN and connect 
2. Block UDP traffic by following these instructions: https://app.asana.com/0/0/1207029295107667/f
3. Open the Console.app and filter by Category: `Network Protection`. The failure recovery logs are indicated with a 🟢 so you can also look for them.
4. After around 5 minutes you will see logs indicating that a register request has been made.
5. It will error with no recovery required as there is not a real problem.

**OS Testing**:


—
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
